fftools/ffmpeg_qsv:Change width and height of hw_frame_ctx to origina…

wenbinc-Bin opened the pull request:
…l value

We use w/h from vaapi hw_frame_ctx to create libva surfaces and we use Param
to init mfx-codec or mfx-vpp. All this value are already aligned value,
so we needn't to aligned width and height in qsv hw_frame_ctx.
We use hw_frame_ctx to create AVFrame, we use hw_frame_ctx->w/h to
configure crop_w/h in hwcontext_qsv.c/qsv_init_surface() function. And
this value should be original width and height, so I change the qsv's
hw_frame_ctx->width/height to original value because of these two reasons.

Signed-off-by: Wenbin Chen wenbin.chen@intel.com

wenbinc-Bin commented:
Hi @xuguangxin, is this what you said that we should change width and height inside hw_frame_ctx?
There is an error about psnr and I think it is caused by aligning a large number for h264 surface. I will find a way to fix it, but I want to make sure that I am on the right direction.

xuguangxin commented:
Could you explain why we need align to 64?
What's happened if encode surface is mapped from 3D and it's not 64 aligned.

thanks

wenbinc-Bin commented:
Hi @xuguangxin :
I tested on different codec and found 64 is the largest aligned number used by MSDK so far, so I align w/h to 64. However I found this affect psnr, and I am trying to use another way to do this.

I think if the mapped surfaces is not same as hw_frame_ctx's configuration, the "hwdownload" section may have potential problem. If the mapped surface is sent directly to encoder, I think it will be ok if the aligned width and height is greater than encoder's need.

xuguangxin commented:
> Hi @xuguangxin :
> I tested on different codec and found 64 is the largest aligned number used by MSDK so far, so I align w/h to 64. However I found this affect psnr, and I am trying to use another way to do this.
> 
> I think if the mapped surfaces is not same as hw_frame_ctx's configuration, the "hwdownload" section may have potential problem. If the mapped surface is sent directly to encoder, I think it will be ok if the aligned width and height is greater than encoder's need.

we'd better find some MediaSDK document for 64, else it's hard to upstream the patch.
Upstream reviewer may ask, why it's not 32? If we said we tested it, reviewer may say, what if msdk changed it in some platforms?

wenbinc-Bin committed:
libavutil/hwcontext_qsv: Change width and height of hw_frame_ctx to original value

The AVFrame created from AVHWFrameContext should have original w/h,
otherwise it is confusing about the real resolution of frame, so I
changed the w/h of hw_frame_ctx to original value.

The w/h will be aligned when we need aligned number, for example, when
we create vaSurface and when we configure mfxSurface. The aligned number
for different codecs are different, so I add two variables into
AVQSVFrameContext to store aligned w/h.

Because I change the w/h of hw_frame_ctx, the avframe used to hwdownload
and hwupload will change too. The qsvvpp need aligned space to store
data, and the AVFrame may do not have sufficient space so I add realloc
operation into hwdownload and hwupload.

Signed-off-by: Wenbin Chen <wenbin.chen@intel.com>

wenbinc-Bin commented:
Hi @xuguangxin. I made changes on the way to aligned w/h. I keep the previous aligned value so we needn't to explain why we use 64.
